docs/architecture/adr-065-store-v2.md

```diff
@@ -0,0 +1,231 @@
+application state. This data structure is the base layer `KVStore`.
+
+In addition, the SDK provides abstractions on top of this Merkle data structure.
+Namely, a root multi-store (RMS) is a collection of each module's `KVStore`.
+Through the RMS, the application can serve queries and provide proofs to clients
+in addition to provide a module access to its own unique `KVStore` though the use
+of `StoreKey`, which is an OCAP primitive.
+
+There are further layers of abstraction that sit between the RMS and the underlying
+IAVL `KVStore`. A `GasKVStore` is responsible for tracking gas IO consumption for
+state machine reads and writes. A `CacheKVStore` is responsible for providing a
+way to cache reads and buffer writes to make state transitions atomic, e.g.
+transaction execution or governance proposal execution.
+
+There are a few critical drawbacks to these layers of abstraction and the overall
+design of storage in the Cosmos SDK:
+
+* Since each module has its own IAVL `KVStore`, commitments are not [atomic](https://github.com/cosmos/cosmos-sdk/issues/14625)
+    * Note, we can still allow modules to have their own IAVL `KVStore`, but the
+      IAVL library will need to support the ability to pass a DB instance as an
+      argument to various IAVL APIs.
+* Since IAVL is responsible for both state storage and commitment, running an 
+  archive node becomes increasingly expensive as disk space grows exponentially.
+* As the size of a network increases, various performance bottlenecks start to
+  emerge in many areas such as query performance, network upgrades, state
+  migrations, and general application performance.
+* Developer UX is poor as it does not allow application developers to experiment
+  with different types of approaches to storage and commitments, along with the
+  complications of many layers of abstractions referenced above.
+
+See the [Storage Discussion](https://github.com/cosmos/cosmos-sdk/discussions/13545) for more information.
+
+## Alternatives
+
+There was a previous attempt to refactor the storage layer described in [ADR-040](./adr-040-storage-and-smt-state-commitments.md).
+However, this approach mainly stems on the short comings of IAVL and various performance
+issues around it. While there was a (partial) implementation of [ADR-040](./adr-040-storage-and-smt-state-commitments.md),
+it was never adopted for a variety of reasons, such as the reliance on using an
+SMT, which was more in a research phase, and some design choices that couldn't
+be fully agreed upon, such as the snap-shotting mechanism that would result in
+massive state bloat.
+
+## Decision
+
+We propose to build upon some of the great ideas introduced in [ADR-040](./adr-040-storage-and-smt-state-commitments.md),
+while being a bit more flexible with the underlying implementations and overall
+less intrusive. Specifically, we propose to:
+
+* Separate the concerns of state commitment (**SC**), needed for consensus, and
```

Comment:
I've been using different type of separation of concerns to guide my design, which is based on the requirement of different roles of clients, it goes like this:
1. consensus state machine (validator node)
  - it requires read/write on the latest version of the state
  - generating the root hash for consensus
  - export/restore state-sync snapshots
2. historical queries without merkle proofs (grpc query server)
  - it requires read-only access to historical states *without* generating merkle proofs
3. historical queries with merkle proofs (`/abci_query?proof=true`)

The **SS** (or the versiondb) is to serve the requirement of 2, I'm experimenting with a new IAVL implementation that will serve the requirement of 1 well, the requirement of 3 is considered as lowest priority, it'll be supported but don't need extreme efficiency.

Reply:
One tricky part is, SC when separated will need duplicate the keys no matter what, which is a significant duplication already, in our evm module, the keys are larger than values (32bytes for value and 20+32 bytes for keys).